message_ix/core.py

```diff
@@ -109,12 +109,27 @@ def add_horizon(scenario, data):
         scenario.add_cat("year", "firstmodelyear", first, is_unique=True)
 
     def vintage_and_active_years(self):
-        """Return a 2-tuple of valid pairs of vintage years and active years for
-        use with data input.
+        """Return a 2-tuple of valid pairs of vintage years and active years
+        for use with data input. A valid year-vintage, year-active pair is
+        one in which:
+
+        - year-vintage <= year-active
+        - both within the model's 'year' set
+        - year-active <= the model's first year
         """
         horizon = self.set('year')
-        combinations = itertools.product(horizon, horizon)
-        year_pairs = [(y_v, y_a) for y_v, y_a in combinations if y_v <= y_a]
+        filters = {'type_year': ['firstmodelyear']}
+        first = (self
```

Comment:
can be written as `self.cat('year', 'firstmodelyear')[0] or self.set('year')[0]`